connector/spanmetricsconnectorv2/connector.go

```diff
@@ -0,0 +1,85 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package spanmetricsconnectorv2 // import "github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2"
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+const scopeName = "otelcol/spanmetricsconnectorv2"
+
+type spanMetrics struct {
+	component.StartFunc
+	component.ShutdownFunc
+
+	next            consumer.Metrics
+	spansMetricDefs map[string]metricDef
+}
+
+func (sm *spanMetrics) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (sm *spanMetrics) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	var multiError error
+	processedMetrics := pmetric.NewMetrics()
+	processedMetrics.ResourceMetrics().EnsureCapacity(td.ResourceSpans().Len())
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		resourceSpan := td.ResourceSpans().At(i)
+		// TODO (lahsivjar): add support for exponential histogram and summary
+		spansHist := newExplicitHistogram(sm.spansMetricDefs)
+
+		for j := 0; j < resourceSpan.ScopeSpans().Len(); j++ {
+			scopeSpan := resourceSpan.ScopeSpans().At(j)
+
+			for k := 0; k < scopeSpan.Spans().Len(); k++ {
+				span := scopeSpan.Spans().At(k)
+				var durationInNs float64
+				startTime := span.StartTimestamp()
+				endTime := span.EndTimestamp()
+				if endTime > startTime {
+					durationInNs = float64(endTime - startTime)
+				}
+				multiError = errors.Join(multiError, spansHist.update(ctx, span.Attributes(), durationInNs))
+			}
+		}
+
+		if len(spansHist.counts) == 0 {
+			continue // don't add an empty resource
+		}
+
+		processedResource := processedMetrics.ResourceMetrics().AppendEmpty()
+		resourceSpan.Resource().Attributes().CopyTo(processedResource.Resource().Attributes())
```

Comment:
Should there be configuration to filter resource attributes, like in spanmetricsconnector?

Reply:
I didn't add it because we didn't need it at the moment. Also, my reasoning for not adding it was because I wanted the connector to only do what is required for span -> metrics production. Other processing (like dropping attributes) if needed can be done by a separate processor.

Reply:
Ah true, we could indeed drop attributes in a subsequent processor since it's stateless. SGTM.